Refactor ChartDownloader

batuakan opened the pull request:
1. Move away from file based caching to MBTiles based caching for better performance and reliability.
    1. Added an option to the webui to vacuum mbtiles to free disk space.
    1. Added an option to the webui to create a separate mbtiles file per region/chart combination.
    1. Added an button under region view to the webui to delete mbtiles files for selected regions/charts.
1. Estimate tile counts for progress reporting without needing to pre-generate tile lists. 
1. Optimize memory usage during tile downloads. Fixed memory leak issue #48 
1. Added possibility to load ChartProvider from a javascript file to be able to calculate dynamic URLs. 
1. Caching of tiles stops if disk usage exceeds a certain threshold. Previously this was only checked for seeding jobs 
1. Added a new endpoint for reporting cache stats for the current session for all chart providers. 
1. Expanded existing cache related endpoints for more flexible usage. 
1. Rewrite of the WebUI, with builtin map using leaflet,
    1. Ability to add/remove/edit regions/polygons/rectangles, and store them on the server side.
    1. Improve the download job management, with better status reporting and error handling.
    1. Added option to web ui to be able refetch tiles from remote server and update local cache. #47 
    1. More user friendly UI
    1. Better support for mobile devices 
1. Various bug fixes and improvements.

Warning
This commit depends on @signalk/mbtiles library. Changes from PR #43 were copied manually

batuakan committed:
Major refactoring of the ChartDownloader code.

Move away from file based caching to MBTiles based caching for better performance and reliability.
    Added an option to the webui to vacuum mbtiles to free disk space.
    Added an option to the webui to create a separate mbtiles file per region/chart combination.
    Added an button under region view to the webui to delete mbtiles files for selected regions/charts.
Estimate tile counts for progress reporting without needing to pre-generate tile lists.
Optimize memory usage during tile downloads.
Added possibility to load ChartProvider from a javascript file to be able to calculate dynamic URLs.
Caching of tiles stops if disk usage exceeds a certain threshold. Previously this was only checked for seeding jobs
Added a new endpoint for reporting cache stats for the current session for all chart providers.
Expanded existing cache related endpoints for more flexible usage.
Rewrite of the WebUI, with builtin map using leaflet,
    Ability to add/remove/edit regions/polygons/rectangles, and store them on the server side.
    Improve the download job management, with better status reporting and error handling.
    Added option to web ui to be able refetch tiles from remote server and update local cache.
Various bug fixes and improvements.

Warning
This commit depends on @signalk/mbtiles library. Changes from PR #43 were copied manually

tkurki commented:
Can you please rebase this on #43 ? Or I can do it for you and you continue from there.

I'd like to move forward so that we can merge #43 (that I just updated) and this and publish a new version.

batuakan commented:
> Can you please rebase this on #43 ? Or I can do it for you and you continue from there.
> 
> I'd like to move forward so that we can merge #43 (that I just updated) and this and publish a new version.

Hi @tkurki 

Can you do it for me, I tried to base my solution on the mbtiles branch after our previous talk but i failed to do so

tkurki committed:
chore: update typescript to support node:sqlite

tkurki committed:
fix: allow undefined onlineChartProviders

tkurki commented:
Rebased and tests passing - you take it from here?

batuakan commented:
> Rebased and tests passing - you take it from here?
@tkurki Nicely done, I couldnt do this earlier because the native mbtiles branch needed to be merged into main before i could rebase my feature branch on it.

I am mostly happy with the code as it is just the modal welcome page/documentation needs to be improved. In the meanwhile can you run my feature branch and see the new features and fixes before we move on to actually reviewing the code. I am excited to hear your feedback on the new features

tkurki commented:
Reload should not take you to the beginning: save chart viewport and the selected visible layers

The modal initial help is less useful than it could be, as when you first see it you don't know what the page is going to look like, so it is hard to understand what it is saying => make the help panel always initially visible at bottom left corner for example, with the ability to minimise (and remaximise afterwards)

Vacuuming is only useful if the mbtiles already exists and you choose to refresh also existing tiles, right? I would change the checkboxes to 
- [ ] use separate file for this region (default: single file per chart provider)
- [x] refresh all tiles
- [ ] compact file after seeding

and refresh all tiles would be gray if there are no existing cached tiles, and compact would be gray unless you select refresh all. Vacuum is not end user friendly terminology....

The jobs button is really low key. It could show an (animated)  indicator if there are jobs running. 

Why can you start a job that is already completed? You can't change its parameters.

Start and stop buttons seem a little unresponsive - like I need to press them multiple times? We could implement a websocket mechanism for sending the status of the jobs instead of polling.

The Serif font in the help panel hurts my brain 😁

torevalley commented:
Installed the branch and have a few questions comments.

First of all great effort and function, I really like the way your regions work especially in edit mode, real easy to e.g. follow a coastline. 

-  I would appreciate the possibility to name the region on creation and edit, not only in seed job creation. I now created my first region and did not name it. I then edited the region and did a new seed, where I named the region Öresund. That did not stick, the mbtiles file and the seedjob still says unnamed --> file is unnamed_eniro.mbtiles.

- Is there any plans to show the existing caches (I guess mbtiles will be the only choice in the near future). I think it is useful to be able to keep inventory and also use that to refresh the content as a seeding job every now and then assuming that the chart providers keep their chart up-to-date.

- I have a chart of type mapstyleJSON that is setup to be cached, it show up as a candidate for seeding job in the "main branch" but on this fork it does not show up. Will that type be supported?

- for the mapstyleJSON a max I use a zoom of 18 or 20 would be really useful (that chart itself has 24 as max but that is overdoing it), realizing quite a few tiles for that but if you are willing to spend the time and the diskspace is there... 
 
- Agree with Teppo on the initial modal, maybe create a default map from the ones already seeded, I might be wrong but I guess most people will create off-line mbtiles mainly from one chart provider, that could eventually be pre-selected.
 

Please let me know if this is not the correct way to submit things, e.g. the first might be a bug with the regions --> maybe better as an issue.....

Hope I can help testing, so let me know of any special directions you want me to exercise :)

tkurki commented:
@batuakan just to be clear - are you planning to address my feedback or do you expect this to be merged as is, then published as beta for testing or a release?

batuakan commented:
Lets get this going again, and sorry responding late, was busy with other SK projects

@tkurki I can add a settings endpoint similar to freeboardsk, to save and load the viewpoint and the zoom and the active regions.

I added a new button on the bottom corner, but the contents of the help dialog is still a matter of procrastination :D

Changed the check boxes, for separate mbtiles creation, I should explicitly say in the documentation that first the tiles are cached into the global tile cache, once that is finished a new mbtiles is created and the tiles are copied to the new one. If you use the layer you want to seed there will definitely be some tiles in the cache for the area you are browsing. 

I changed the icon on the jobs button, I can look into how to add a badge to the jobs button to show the number of ongoing jobs.

I will remove the start/stop button from the jobs view then, the user can cancel and remove a job by using the X button at the top right corner.

The buttons are bit unresponsive depending on when you click because at every poll to the server, the whole jobs view is discarded and recreated, but once we remove the buttons the situation should improve

Any font recommendations?

I don't think you would approve to merge this code before going through a strict code review :D. I am expecting many rounds of back and forth

batuakan commented:
@torevalley 

I am happy it worked for you.

- Region settings dialog pops up when you finish creating a region, I think it is the best time to name the region

- I wanted to have a feature like that, but i couldn't think of a good way to visualize that. Also if you can see the tile when you are online, then you have it in the cache, if you dont see the tile when you are offline then you dont have it in the cache :D.

- regarding the mapstyleJSON, apparently I filter out everything else other than tilelayer. Did the caching work nicely with mapstyleJSON? I wonder if it would cache into mbtiles file?

torevalley commented:
About regions, I understand what you are saying but changing the name in edit still makes sense.
I created an unnamed region and it seems it cannot be named after creation.

On visualize regions, does the cache hold info on the region it was created from?
If the region is available and time for latest seed on that area could be stored in my head that could be visualized.
Another way could be to just persist the seed jobs per region and make if possible to display that list. Thinking that as a user you would know your regions.

Mapstyle json did not cache as tiles, but you can add it and check the proxy box. I have no clue whats happening behind the scenes with that type of source/provider. I was merely hoping it would be cachable :)

batuakan committed:
- Updated icons
- Removed start/stop button under seeding jobs view
- Added a badge showing number of active jobs
- Added a new button to toggle help modal
- Updated the checkboxes text in the region modal view

batuakan committed:
Added more comprehensive help

batuakan commented:
Hi @tkurki 

I added the help page as well, shall we begin the code review process?

batuakan committed:
Fixed an issue with charts app with null settings

batuakan committed:
Minor code fix

batuakan committed:
Minor code fix regarding chart plugin webapp settings

batuakan commented:
@tkurki Shall we begin the cod review processes for this PR?

tkurki commented:
Apologies, i have just too much stuff on my plate. Maybe @msallin could help here?

msallin reviewed (commented):
Looked at this end to end. Blockers and security first; the rest is incremental.

## Blockers

**1. `getTilesForBBox` ignores its `minZoom` parameter** — `src/chartDownloaderTileHelpers.ts:30,39`, inner loop starts at `z = 0`. Also imported but never called from `chartDownloader.ts:9`. Fix or delete.

**2. `deleteCache()` throws on bbox-initialized jobs** — `tilesInDB` is only assigned in `initalizeJobFromFeature` (`chartDownloader.ts:123`); `initializeJobFromBBox` at line 139 never sets it, and line 262 calls it unconditionally, so it hits a `TypeError`. Reachable via `POST /cache/jobs/:id` with `action='delete'`.

**3. `POST /cache/:identifier` mis-routes actions and lies about state** — `src/index.ts:400-410`:
```ts
if (action === 'start' || bbox) { job.seedCache() }
else if (action === 'delete') { job.deleteCache() }
```
`action='delete' + bbox` silently becomes a seed. `seedCache()` is fire-and-forget; the handler returns `state: 'COMPLETED'` while the job is actually just starting. Line 376 also returns `500` for missing provider, should be `404`.

**4. `CachingDisabled` is a one-way latch** — `chartDownloader.ts:353` sets it, nothing ever clears it. Survives for the process lifetime; caching does not resume when disk space frees up.

**5. `totalTiles` auto-inflates mid-run** — `chartDownloader.ts:213-219` bumps the total upward when running count exceeds the estimate. Progress bar visibly stalls near 100% and then "restarts". Tighten the estimator or always walk `countTiles` with a higher cap.

**6. `seedCache()` transitions `state = Stopped` from inside a running worker** — line 192 fires while sibling workers in the same `Promise.all` are still executing. Let the post-`Promise.all` block at line 228 do the single transition.

## Security

**7. Path traversal via region name** — user-controlled `feature.properties.name` flows unsanitized from `public/js/index.js:496` through `chartDownloader.ts:135` to the filename at line 236 (`${this.regionName}_${identifier}.mbtiles`). `path.join` does not strip `..`. Sanitize to `[A-Za-z0-9_\- ]+`, or `path.resolve` the result and verify it is under the cache root.

**8. `.js` chart providers `import()`-ed at startup** — `src/charts.ts:43-47` executes any `.js` in a chart directory with full Node privileges. At minimum a prominent README warning; an allowlist would be better.

## Earlier-discussion items still open

**9. Region rename workflow bug** (raised by @torevalley). Confirmed: Save and "Create Seeding Job" are separate buttons. `createJob` at `public/js/index.js:430` reads `activeRegionLayer.toGeoJSON()` without first copying `regionName.value` into `feature.properties.name`, so a rename typed into the form is lost unless Save is clicked first, which explains the `Unnamed_*.mbtiles` file @torevalley hit. Fix: copy the name inside `createJob`, or merge the two buttons.

**10. `mapstyleJSON` providers filtered out of seeding** (raised by @torevalley). `public/js/index.js:208` hardcodes `info.type === 'tilelayer'`. Drop the filter or document that style-json providers are not seedable by design.

**11. "Restart completed job"** (raised by @tkurki). `src/index.ts:429` still accepts `action='start'` regardless of current state. UI hides the button, but the endpoint is permissive. Reject or document.

**12. Checkbox dependency greying** (raised by @tkurki). `public/index.html:110,117` — no disabled logic. "Refresh all tiles" should grey when no cached tiles exist; "compact" should grey unless refresh is checked.

**13. Jobs polling** (raised by @tkurki). Still `setInterval(fetchActiveJobs, 2000)` at `public/js/index.js:254` and the DOM is fully recreated on every poll. Either diff-render, or move to a websocket.

## Non-blocking cleanup

- `_mbtilesHandle._db` private-attribute coupling at `chartDownloader.ts:125,250,261,270,281` and `mbtiles-helpers.ts:41-48`. Worth raising with `@signalk/mbtiles`.
- `plugin.stop()` at `index.ts:206` does not cancel `ActiveJobs`, close handles, or reset static state. In-place restart leaks.
- `hasDiskSpace` at `chartDownloader.ts:477` treats any error as "no space", which latches `CachingDisabled` on transient failures.
- `db.prepare` inside loops: hoist out of the zoom loop (`mbtiles-helpers.ts:95`) and the chunk loop (`mbtiles-helpers.ts:150`); add `ROLLBACK` around the chunk transaction.
- Worker and `getTileFromCacheOrRemote` both independently poll disk space (`chartDownloader.ts:189` and `350`) with different semantics (cancel-job vs disable-caching). Consolidate or at least document.
- `any` on the mbtiles handle: one `type MBTilesHandle = { _db: DatabaseSync; getTile; putTile; startWriting; putInfo }` eliminates all seven `no-explicit-any` disables.
- Magic numeric enums crossing the API boundary: `job.type === 1|2`, `item.state === 1` at `public/js/index.js:338,346,410`. Mirror the enum or return strings from `info()`.
- Duplicate `toggleHelpModal` at `public/js/index.js:260` targets a non-existent `helpModal`; only the definition at line 610 works.
- Implicit globals: `overlayMaps`, `layer`, bare `regionName` at lines 200, 214, 496, 599.
- Help modal renders as a full-screen blocking overlay on first load (`public/index.html:134` missing `hidden`; `.modal` is full-screen fixed). Does not match @tkurki's floating-panel ask.
- Dead code/deps: commented start/stop blocks (`public/js/index.js:284,357`, `public/index.html:73,95,120`); vendored `jquery-3.5.1.min.js` with zero usage; `p-limit` still in `package.json` but no longer imported.
- Error-string control flow: `err.message?.includes('Tile does not exist')` at `chartDownloader.ts:388` and `index.ts:658`.
- `concurrency = 32` magic number (was `concurrentDownloadsLimit = 20`). Promote to a named const with a reason.
- `.mbtiles_` trailing underscore at `index.ts:287`. Intentional to avoid collision with user charts? Add a comment.
- `convertBboxToGeoJSON` uses clockwise winding (`tile-helpers.ts:171`) while the sibling `bboxPolygon` at line 262 is counter-clockwise. turf tolerates it.
- `console.log(settings)` left in at `public/js/index.js:45`.

## Tests

No new tests for ~1500 lines of new TS. Minimum: unit tests for `chartDownloaderTileHelpers` (coordinate math, antimeridian split, estimator on known shapes), in-memory `node:sqlite` tests for the MBTiles helpers, smoke tests for the new `/cache/*` endpoints. The existing `chai-http` setup is reusable.

---

Happy to pair on the blockers. Suggest landing 1–8 first and rolling 9–13 plus the cleanup list into follow-up commits on the same branch.

batuakan commented:
> @batuakan does that also incorporate #54 and #59?

I think those PR's should be merged first, and i will rebase this PR, because I have moved those functions into their helper files

msallin commented:
@batuakan okay, I bring them in..!

msallin commented:
@batuakan I can Claude (Opus 4.7, 1M, max) do the rebase and commit to your branch so that you can review. Should do the heavy lifting for you. Then, at the end, we squash all the commits before merge. Ok?

I will do that after I got some more hygienic PRs in.

msallin commented:
@batuakan to wrap up: I opened [#84](https://github.com/SignalK/charts-plugin/pull/84) with your branch rebased onto master (v3.6.0) and squashed into one commit authored by you. Your `feature/mbtiles` branch is untouched. Continue review and any further work on #84; this PR (#49) can stay open for context or be closed in favour of #84 — your call.

msallin commented:
Decided to work with @batuakan on new PR.
